pkg/applyset/applyset.go

```diff
@@ -0,0 +1,543 @@
+
+	if err := a.desired.Add(obj); err != nil {
+		return err
+	}
+	// If a load callback is provided, we will load the object from the cluster.
+	if a.loadFn != nil {
+		if err := a.loadFn(obj, observed); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// ID is the label value that we are using to identify this applyset.
+// Format: base64(sha256(<name>.<namespace>.<kind>.<apiVersion>)), using the URL safe encoding of RFC4648.
+func (a *applySet) ID() string {
+	unencoded := strings.Join([]string{
+		a.parent.GetName(),
+		a.parent.GetNamespace(),
+		a.parent.GetKind(),
+		a.parent.GroupVersionKind().Group,
+	}, ApplySetIDPartDelimiter)
+	hashed := sha256.Sum256([]byte(unencoded))
+	b64 := base64.RawURLEncoding.EncodeToString(hashed[:])
+	// Label values must start and end with alphanumeric values, so add a known-safe prefix and suffix.
+	return fmt.Sprintf(V1ApplySetIdFormat, b64)
+}
+
+func (a *applySet) injectToolLabels(labels map[string]string) map[string]string {
+	if labels == nil {
+		labels = make(map[string]string)
+	}
+	if a.toolLabels != nil {
+		for k, v := range a.toolLabels {
+			labels[k] = v
+		}
+	}
+	return labels
+}
+
+func (a *applySet) InjectApplysetLabels(labels map[string]string) map[string]string {
+	if labels == nil {
+		labels = make(map[string]string)
+	}
+	labels[ApplysetPartOfLabel] = a.ID()
+	return labels
+}
+
+type applySetUpdateMode string
+
+var updateToLatestSet applySetUpdateMode = "latest"
+var updateToSuperset applySetUpdateMode = "superset"
+
+// updateParentLabelsAndAnnotations updates the parent labels and annotations.
+func (a *applySet) updateParentLabelsAndAnnotations(ctx context.Context, mode applySetUpdateMode) (sets.Set[string],
+	sets.Set[string], error) {
+	original, err := meta.Accessor(a.parent.DeepCopyObject())
+	if err != nil {
+		return nil, nil, err
+	}
+
+	parentPatch := &unstructured.Unstructured{}
+	parentPatch.SetUnstructuredContent(map[string]interface{}{
+		"apiVersion": a.parent.GetAPIVersion(),
+		"kind":       a.parent.GetKind(),
+		"metadata": map[string]interface{}{
+			"name":      a.parent.GetName(),
+			"namespace": a.parent.GetNamespace(),
+		},
+	})
+	// Generate and append the desired labels to the parent labels
+	desiredLabels := a.desiredParentLabels()
+	labels := a.parent.GetLabels()
+	if labels == nil {
+		labels = make(map[string]string)
+	}
+	for k, v := range desiredLabels {
+		labels[k] = v
+	}
+	parentPatch.SetLabels(labels)
+
+	// Get the desired annotations and append them to the parent
+	desiredAnnotations, returnNamespaces, returnGKs := a.desiredParentAnnotations(mode == updateToSuperset)
+	annotations := a.parent.GetAnnotations()
+	if annotations == nil {
+		annotations = make(map[string]string)
+	}
+	for k, v := range desiredAnnotations {
+		annotations[k] = v
+	}
+	parentPatch.SetAnnotations(annotations)
+
+	options := metav1.ApplyOptions{
+		FieldManager: a.fieldManager + "parent-labeller",
```

Comment:
is the field-manager set with a prefix already? otherwise this manager will look very strange

Reply:
done.